### PR TITLE
Add an API to get the `TypeName` for a `TypeReference`.

### PR DESCRIPTION
**Pull request by vRallev**

### CODE REVIEW

File: compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/KotlinPoetUtils.kt [hunk 1]
```diff
@@ -229,6 +233,10 @@ public fun KotlinType.asTypeName(): TypeName {
  *                      resolve type arguments.
  */
 @ExperimentalAnvilApi
+@Suppress("DeprecatedCallableAddReplaceWith")
+@Deprecated(
+  "Don't rely on descriptors and make the code agnostic to the underlying implementation."
+)
 public fun KotlinType.asTypeNameOrNull(
```
Comment (JoelWilcox) [inline compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/KotlinPoetUtils.kt:240]: I noticed this one mentions descriptors but the deprecation message for `KotlinType.asTypeName()` mentions PSI. Should one be swapped to match?

File: compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/reference/TypeReference.kt [hunk 2]
```diff
@@ -52,12 +70,38 @@ public sealed class TypeReference {
         ?.toClassReference(module)
     }
 
+    override val typeName: TypeName? by lazy(NONE) {
+      try {
+        type.requireTypeName(module)
+      } catch (e: AnvilCompilationException) {
+        // The goal is to inline the function above and then stop throwing an exception.
```
Comment (JoelWilcox) [inline compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/reference/TypeReference.kt:77]: May be worth adding a TODO here just so it's obvious this isn't final when scanning the code

File: compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/reference/TypeReference.kt [hunk 3]
```diff
@@ -52,12 +70,38 @@ public sealed class TypeReference {
...
+        typeReference.requireTypeName(module)
+          .let { typeName ->
+            // This is a special case when mixing parsing between descriptors and PSI.  Descriptors
+            // will always represent lambda types as kotlin.Function* types, so lambda arguments
+            // must be converted or their signatures won't match.
+            // see https://github.com/square/anvil/issues/400
+            (typeName as? LambdaTypeName)?.asFunctionType() ?: typeName
+          }
+      } catch (e: AnvilCompilationException) {
+        // The goal is to inline the function above and then stop throwing an exception.
```
Comment (JoelWilcox) [inline compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/reference/TypeReference.kt:101]: Same here for adding TODO

**Review by JoelWilcox — APPROVED** [review]
All super minor comments

File: compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/reference/TypeReference.kt (hunk 3 above)
Comment (vRallev) [reply compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/reference/TypeReference.kt:101]: I removed one of these methods already in the next PR. I'll add a TODO to the remaining method in the other stacked PR.

**Pull request merged**